idr/marshal2.go

```diff
@@ -0,0 +1,205 @@
+			break
+		}
+		// attribute child node is ignored.
+	}
+	return textNodeFound && !elemNodeFound
+}
+
+func isChildArray(n *Node) bool {
+	if IsJSONArr(n) {
+		return true
+	}
+	// CSV or fixed-length formats won't produce array.
+	//
+	// JSONStreamReader, when encountered actual array in JSON input, produces
+	// a collection of child element nodes with their Data fields all equal to
+	// empty string.
+	//
+	// XML/EDI are a bit more complicated:
+	// There is not concept of array in XML, neither in EDI. XML, however, can simulate
+	// array by having a number of child elements with the same element name. EDI has
+	// segment loop. So the primary idea is to check child element nodes' names and use
+	// their similarity to determine whether the children represent an array or not.
+	// Throwing a wrench here is that fact in XML, there are dummy text nodes sprinkled
+	// around, in between XML elements, so we need to ignore them. Also there could be
+	// attribute nodes in XML, need to ignore them too.
+	//
+	// One ambiguous case in XML (and similarly in EDI):
+	//   <abc>
+	//      <child>something</child>
+	//   </abc>
+	// We can consider node 'abc' contains a field 'child', or we can consider node 'abc'
+	// contains an array of one element which is 'child'. There is really no other signs/flags
+	// help us to disambiguate the situation. So use common sense here, treat this situation
+	// like a single child field, instead of a child array with one element. EDI is similar.
+	// EDI doesn't have some indicator (i.e. a segment can have multiple instances), but it's
+	// too hard to pass the indicator here. So hope this promise/limitation is acceptable.
+	elemNum := 0
+	elemName := (*string)(nil)
+	for c := n.FirstChild; c != nil; c = c.NextSibling {
+		if c.Type != ElementNode {
+			continue
+		}
+		elemNum++
+		if elemName == nil {
+			elemName = strs.StrPtr(j2NodeName(c))
+		} else if j2NodeName(c) != *elemName {
+			// We have differently named child element nodes, clearly
+			// this isn't an array case.
+			return false
+		}
+	}
+	// If the number of identically named child element nodes is > 1 (including JSON's "" named child
+	// element nodes) then it's an array case;
+	// If there is only one child element node, then it is array case only when its name is "" - that's
+	// in JSON array case. Otherwise, just treat it as a single child field case.
+	return elemNum > 1 || (elemNum == 1 && *elemName == "")
+}
+
+func getChildText(n *Node) interface{} {
+	if !IsJSON(n) {
+		return n.InnerText()
+	}
+	n = n.FirstChild
+	switch {
+	case IsJSONValueNum(n):
+		f, _ := strconv.ParseFloat(n.Data, 64)
+		return f
+	case IsJSONValueBool(n):
+		b, _ := strconv.ParseBool(n.Data)
+		return b
+	case IsJSONValueNull(n):
+		return nil
+	default:
+		return n.Data
+	}
+}
+
+// J2NodeToInterface converts *Node into an interface{} that, once marshaled, is JSON friendly.
+func J2NodeToInterface(n *Node) interface{} {
```

Comment:
looks like we want to convert xml to json? why we want do this?

Reply:
There was demand that people wanted to simple copy a section of json input to output, so we introduced this JSONify2 (under a different name) . However the code was written such that it works for all Node trees from different input formats.

Not only that, now we have `javascript` custom function, which we pass in a special argument `_node` that is JSON serialized Node of the current context, so schema writer can do whatever custom processing they see to fit. In this case it's even more important to pass in a json doc translated from all supported input formats (not just JSON/XML, but all input formats)